typing/typecore.ml

```diff
@@ -2874,14 +2894,38 @@ let collect_unknown_apply_args env funct ty_fun0 rev_args sargs =
               let ty_arg = newvar () in
               let ty_res = newvar () in
               if get_level ty_fun >= get_level ty_arg &&
-                 not (is_prim ~name:"%identity" funct)
+                not (is_prim ~name:"%identity" funct)
               then
                 Location.prerr_warning sarg.pexp_loc
                   Warnings.Ignored_extra_argument;
               unify env ty_fun (newty (Tarrow(lbl,ty_arg,ty_res,commu_var ())));
               (ty_arg, ty_res)
           | Tarrow (l, ty_arg, ty_res, _) when labels_match ~param:l ~arg:lbl ->
               (ty_arg, ty_res)
+          | Tfunctor (l, _, (p, fl), _) when l = lbl
+            || !Clflags.classic && lbl = Nolabel && not (is_optional l) ->
+              begin match unify_to_arrow env ty_fun with
```

Comment:
Decaying `Tfunctor` to `Tarrow` as soon as the type of the function is not principally known feels a bit harsh:
```ocaml
module type S = sig type t val sum: t -> t -> t end
module I = struct type t = int let sum = (+) end
let sum (module X:S) x y = X.sum x y
let err = Fun.id sum (module I:S) 0 1
```
fails with
>```ocaml
>Error: The value sum has type (module X : S) -> X.t -> X.t -> X.t
>       but an expression was expected of type (module S) -> 'a -> 'b -> 'c
>       The module X would escape its scope
>```
 whereas
```ocaml
let ok = (Fun.id sum) (module I:S) 0 1 
```
works.